[PyTorch] Support borrowing in/out Tensors in TensorIterator

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55691 [PyTorch] Migrate add operators to borrow in TensorIteratorBase
* **#55690 [PyTorch] Support borrowing in/out Tensors in TensorIterator**
* #55940 [PyTorch] Always update op.current_dtype in TensorIteratorBase::set_output
* #55420 [PyTorch] Remove non-const TensorIterator::tensor() method
* #55258 [PyTorch] Format generated structured kernels code better
* #55351 [PyTorch] Fix const correctness for resize native functions

Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

swolchok committed:
[PyTorch] Support borrowing in/out Tensors in TensorIterator

Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

ezyang reviewed (approved):
This is fine; but for other readers note we need to resolve some structured kernel interactions in the previous PR

swolchok committed:
rebase on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
fix newly-added resize_as_sparse_ on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
rebase w/conflicts on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
fix build at bottom of stack on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Support borrowing in/out Tensors in TensorIterator"


Just change `OperandInfo::tensor` and
`TensorIteratorConfig::tensors` to hold `c10::MaybeOwned<Tensor>`, and
deal with the consequent pointer syntax. Had to C10_ALWAYS_INLINE
OperandInfo to preserve existing inlining behavior for whatever
compiler-idiosyncratic reason.

This is a separate diff from usage to enable measuring the cost of
support, and because there is no reason not to send it separately.

We probably should not land this without a plan to migrate a lot of
TensorIterator use cases to use either borrowing or structured kernels
+ borrowing.

Differential Revision: [D27607293](https://our.internmc.facebook.com/intern/diff/D27607293/)

[ghstack-poisoned]
